### DISCUSSION TITLE
Graphql workspaces

### DISCUSSION THREAD

**Discussion by joselcvarela** [directus/directus#5362] [Archive]
Now we have `/system` and `/` to work with data from directus system and from our model system, respectively.

Although, as I mentioned in Discord, it makes more difficult for consumers to maintain two clients.
Moreover, if we had `user_created` in our model system, we cannot query this field in `/` workspace, because it belongs to `/system`.
As workaround I will need to fetch the id from `/` and then request `/system`.

In my point of view, this takes off the versatility of GraphQL.

The reason to separate between this two workspaces is to avoid naming conflicts. But I think we could use another name convention on the fields, for example, the field `create_users_item` could be:
-  `create_directus_users_item`
-  `create_system_users_item`
-  `directus_create_users_item`
-  `system_create_users_item`

Is there any reason to not follow this approach?

**Comment by rijkvanzanten** [comment]
> Moreover, if we had user_created in our model system, we cannot query this field in / workspace, because it belongs to /system.

This is not true, as the schema is still the same across both endpoints. The only difference is the fields that are available on the root Query / Mutation.

> Is there any reason to not follow this approach?

Yes, it's partly white-listing reasons (preventing the name `directus` in IO usage), but more importantly: these aren't the only resolvers/mutations available. There's other query resolvers for things like server health, extensions, and mutations for things like login, tea, hashes etc etc. We can't reliably prevent naming conflicts between these and user-created names. 


---

At the end of the day, this is really just a limitation in what GraphQL can do.. We started off with everything combined, where the user collections where "scoped" under an `items` key to prevent this, but that was giving trouble with existing GraphQL tooling like Apollo Cache (see https://github.com/directus/directus/discussions/4265 for a lengthy discussion around that). I wish there was just a native way in GraphQL to scope groups of resolvers like that, almost like

```graphql
QuerySystem {}
// vs
Query {}
```

or

```graphql
Query on System {
}
```

but as of right now, you can only have a single root Query object, so it seems like this double endpoint structure seems like the next-best way to scope these resolvers 🤷🏻

**Comment by joselcvarela** [comment]
I'm sad 😢but I understand and agree with you.

> > Moreover, if we had user_created in our model system, we cannot query this field in / workspace, because it belongs to /system.

> This is not true, as the schema is still the same across both endpoints. The only difference is the fields that are available on the root Query / Mutation.

I am querying `id` field from `user_created` on `/` but the response is always `user_created: null`
Am I missing something?

**Reply by rijkvanzanten** [reply]
Not sure! Could you be a separate bug / wrong query / maybe the user doesn't exist / field isn't set
